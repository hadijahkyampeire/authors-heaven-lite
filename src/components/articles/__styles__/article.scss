.article-page {
  margin: auto;
  width: 80%;

  .article-title {
    margin-bottom: 1rem;
    font-weight: bold;
    color: rgba(41, 41, 41, 1);
    font-size: 2rem;
  }
  .other-header {
    display: flex;
    margin-bottom: 2rem;

    .author {
      margin-right: 0.5rem;
      text-transform: capitalize;
      color: black;
      font-weight: bold;
    }
  }
  .header {
    display: flex;
    align-items: center;
    margin-bottom: 2rem;
    a {
      margin-right: 1rem;
    }
    h2 {
      font-size: 1.5rem;
    }
  }
  .article-footer {
    display: flex;
    margin-top: 2rem;
    border-top: 1px solid lightgray;
    border-bottom: 1px solid lightgray;
    padding: 0.5rem 0;
    button {
      margin-right: 0.5rem;
      border: none;
      outline: none;
      padding: 0.5rem;
    }
  }
}
