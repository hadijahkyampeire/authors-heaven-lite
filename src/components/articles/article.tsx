import { useLocation } from "react-router";
import { Link } from 'react-router-dom';

import { getSingleArticle, deleteArticle } from 'actions/articles';
import React, { useEffect, useState } from 'react';
import { ApiArticle } from 'types/articles';
import { Header, DialogModal } from "components/commons";

import './__styles__/article.scss';

interface Props {
  article?: ApiArticle;
  fetchArticle: (...args: Parameters<typeof getSingleArticle>) => void;
  deleteArticle: (...args: Parameters<typeof deleteArticle>) => void;
  email?: string;
};

export const Article: React.FC<Props> = ({ article, fetchArticle, email, deleteArticle }) => {
  const [openDelete, setOpenDelete] = useState<string | undefined>(undefined);
  const { pathname: url } = useLocation();

  useEffect(() => {
    fetchArticle(url);
  }, [url]); // eslint-disable-line react-hooks/exhaustive-deps

  const hasPermission = email === article?.author?.email;
  return (
    <section className="article-page">
      <Header title={`Viewing ${article?.title }`}/>
      <div>
        <div className="article-title">{article?.title}</div>
        <div className="other-header"><span className="author">By: {article?.author?.username}</span><span>{new Date().getDate()}</span><span>2 min read</span></div>
        <div className="article-body">
          <div className="article-description">{article?.description}</div>
          <div className="article-body">{article?.body}</div>
        </div>
        {hasPermission ? (
          <>
            <div className="article-footer">
              <Link to={`${url}/edit`}>Edit Article</Link>
              <button onClick={() => setOpenDelete(article?.slug)}>Delete Article</button>
              <button>Share Article</button>
            </div>
          </>
          )
        : null}
        <DialogModal 
          open={openDelete === article?.slug} 
          closeModal={() => setOpenDelete(undefined)}
          onSubmit={() => deleteArticle(article?.slug)}
          title={`Delete ${article?.title}`} 
          text='Are you sure you want to delete this Article?' 
           />
      </div>
    </section>
  );
};
