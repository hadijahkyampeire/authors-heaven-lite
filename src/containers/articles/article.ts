import { connect } from 'react-redux';

import { getSingleArticle } from 'actions/articles';
import { Article } from 'components/articles';
import { AppState } from 'types';


const mapStateToProps = (state: AppState) => ({
  article: state.articles?.article
});
const mapDispatchToProps = { fetchArticle: getSingleArticle };

export const ArticlePage = connect(
  mapStateToProps,
  mapDispatchToProps
)(Article);
