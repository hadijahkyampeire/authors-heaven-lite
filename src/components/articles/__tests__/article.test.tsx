import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { fireEvent } from "@testing-library/react";
import { render } from "testUtils";

import { Article } from '../article';

type ArticleProps = React.ComponentProps<typeof Article>;

const baseProps: ArticleProps = {
  fetchArticle: jest.fn(),
  deleteArticle: jest.fn(),
  article: { title: '1 one', description: 'desc', body: 'body', slug: '1-one', author: { username: 'test user', email: 'test@gmail.com' }}
};

function renderUI(props: Partial<ArticleProps> = {}) {
  return render(
    <Router>
      <Article {...baseProps} {...props} />
    </Router>
  );
};

describe('<Article />', () => {
  it("snapshot test", () => {
    const { container } = renderUI({ email: 'test@gmail.com'});
    expect(container).toMatchSnapshot();
  });

  it('renders properly', async () => {
    const { findByTestId, getByText } = renderUI({  email: 'test@gmail.com' });

    const articlePage = await findByTestId('article-page');
    expect(articlePage).toBeInTheDocument();
    expect(getByText('Viewing 1 one')).toBeInTheDocument();
  });
  
  it('renders footer buttons if you are the right user/handles click on yes', async () => {
    const { findByTestId, getByText, queryByTestId } = renderUI({  email: 'test@gmail.com' });
    const articleFooter = await findByTestId('article-footer');
    expect(articleFooter).toBeInTheDocument();

    expect(queryByTestId('dialog-modal')).toBeNull();

    const deleteButton = getByText('Delete Article');
    fireEvent.click(deleteButton);
 
    const deleteModal = await findByTestId('dialog-modal');
    expect(deleteModal).toBeInTheDocument();

    const yesButton = getByText('Yes');
    fireEvent.click(yesButton);
    expect(queryByTestId('dialog-modal')).toBeNull();
  });

  it('handles close button', async () => {
    const { findByTestId, getByText, queryByTestId } = renderUI({  email: 'test@gmail.com' });
    const articleFooter = await findByTestId('article-footer');
    expect(articleFooter).toBeInTheDocument();

    const deleteButton = getByText('Delete Article');
    fireEvent.click(deleteButton);
 
    const deleteModal = await findByTestId('dialog-modal');
    expect(deleteModal).toBeInTheDocument();

    const cancelButton = getByText('No');
    fireEvent.click(cancelButton);
    expect(queryByTestId('dialog-modal')).toBeNull();
  });

  it('Does not render footer buttons if you are not the right user', async () => {
    const { queryByTestId } = renderUI({  email: 'notuser@gmail.com' });
   
    expect(queryByTestId('article-footer')).toBeNull();
  });
});
