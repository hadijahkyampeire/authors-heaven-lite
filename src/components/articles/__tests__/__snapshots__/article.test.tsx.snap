// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> snapshot test 1`] = `
<div>
  <section
    class="article-page"
    data-testid="article-page"
  >
    <div
      class="app-header"
    >
      <span
        class="bx--tooltip--definition bx--tooltip--a11y"
      >
        <button
          aria-describedby="definition-tooltip-1"
          class="bx--tooltip__trigger bx--tooltip--a11y bx--tooltip__trigger--definition bx--tooltip--top bx--tooltip--align-start"
          type="button"
        >
          <div>
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="32"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M14 26L15.41 24.59 7.83 17 28 17 28 15 7.83 15 15.41 7.41 14 6 4 16 14 26z"
              />
            </svg>
          </div>
        </button>
        <span
          class="bx--assistive-text"
          id="definition-tooltip-1"
          role="tooltip"
        >
          Go Back
        </span>
      </span>
      <div
        class="title"
      >
        Viewing 1 one
      </div>
    </div>
    <div>
      <div
        class="article-title"
      >
        1 one
      </div>
      <div
        class="other-header"
      >
        <span
          class="author"
        >
          By: 
          test user
        </span>
        <span>
          20
        </span>
        <span>
          2 min read
        </span>
      </div>
      <div
        class="article-body"
      >
        <div
          class="article-description"
        >
          desc
        </div>
        <div
          class="article-body"
        >
          body
        </div>
      </div>
      <div
        class="article-footer"
        data-testid="article-footer"
      >
        <a
          href="//edit"
        >
          Edit Article
        </a>
        <button>
          Delete Article
        </button>
        <button>
          Share Article
        </button>
      </div>
    </div>
  </section>
</div>
`;
